circe-display-images.el

```diff
@@ -0,0 +1,197 @@
+  ;; https://emacs.stackexchange.com/a/566
+  (interactive)
+  (let*
+      ((inhibit-read-only t)
+       (url (url-get-url-at-point))
+       (image-data(gethash url circe-display-images-text-property-map))
+       (display-image-p (plist-get image-data :display-image-p))
+       (image-property-of-url (plist-get image-data :image-property))
+       (from (if display-image-p 'display 'display-backup))
+       (to (if display-image-p 'display-backup 'display))
+       (current-pos (point-min))
+       left current-image-property)
+    (while (and current-pos (/= current-pos (point-max)))
+      ;; Find the next image property in the buffer.
+      (if (get-text-property current-pos from)
+          (setq left current-pos)
+        (setq left (next-single-property-change current-pos from)))
+      (if (or (null left) (= left (point-max)))
+          (setq current-pos nil)
+        (setq current-image-property (get-text-property left from))
+        (setq current-pos (or (next-single-property-change left from)
+                              (point-max)))
+        ;; Swap the images if our current image matches the image from the URL.
+        (when (equal image-property-of-url current-image-property)
+          (add-text-properties
+           left current-pos (list from nil to current-image-property)))))
+    ;; Make sure to invert the :display-image-p flag after processing all
+    ;; images.
+    (puthash url `(:image-property ,image-property-of-url
+                   :display-image-p ,(not display-image-p))
+             circe-display-images-text-property-map)))
+
+(defun circe-display-images-insert-image-from-url (url)
+  "Attempt to download the image from URL, and insert it."
+  (let ((buffer (url-retrieve-synchronously url)))
+    (when buffer
+      (unwind-protect
+          (let* ((data (with-current-buffer buffer
+                         (goto-char (point-min))
+                         (search-forward "\n\n")
+                         (buffer-substring (point) (point-max))))
+                 (img (create-image
+                       data 'imagemagick t
+                       :max-height circe-display-images-max-height
+                       :background circe-display-images-background)))
+            (when img
+              (insert-image img)
+              ;; Store the image so that we can toggle it on and off later. We
+              ;; know the image is 1 behind us, since we just inserted it.
+              (let* ((image-property
+                      (get-text-property (- (point) 1) 'display)))
+                (puthash url
+                         `(:image-property ,image-property :display-image-p t)
+                         circe-display-images-text-property-map))
+              ;; This is safely a no-op if the image isn't a gif.
+              (when circe-display-images-animate-gifs
+                (image-animate img)))))
+      (kill-buffer buffer))))
```

Comment:
This `unwind-protect` has no clean-up handler because this form is still part of its body.